temporal/tracing.go

```diff
@@ -1,11 +1,136 @@
+	if raw, ok := reader.Get(tracingKey); ok {
+		var kvp map[string]string
+		if err := converter.GetDefaultDataConverter().FromPayload(raw, &kvp); err != nil {
+			return ctx, nil
+		}
+		ctx = context.WithValue(ctx, workflowTraceParametersKey{}, &kvp)
+	}
+	return ctx, nil
+}
+
+func (w workflowTraceParametersPropagator) InjectFromWorkflow(ctx workflow.Context, writer workflow.HeaderWriter) error {
+	kvp, ok := ctx.Value(workflowTraceParametersKey{}).(*map[string]string)
+	if ok {
+		payload, err := converter.GetDefaultDataConverter().ToPayload(*kvp)
+		if err != nil {
+			return err
+		}
+		writer.Set(tracingKey, payload)
+	}
+	return nil
+}
+
+func (w workflowTraceParametersPropagator) ExtractToWorkflow(ctx workflow.Context, reader workflow.HeaderReader) (workflow.Context, error) {
+	if raw, ok := reader.Get(tracingKey); ok {
+		var kvps map[string]string
+		if err := converter.GetDefaultDataConverter().FromPayload(raw, &kvps); err != nil {
+			return ctx, nil
+		}
+		ctx = workflow.WithValue(ctx, workflowTraceParametersKey{}, &kvps)
+	}
+	return ctx, nil
+}
+
 func newOtelInterceptor() (interceptor.Interceptor, error) {
 	// Uses global tracer provider.
 	return temporalotel.NewTracingInterceptor(temporalotel.TracerOptions{})
 }
+
+func NewWorkflowTraceParametersPropagator() workflow.ContextPropagator {
+	return &workflowTraceParametersPropagator{}
+}
+
+func WithTraceParameters(ctx context.Context, entries ...string) context.Context {
+	container := getOrCreateTags(ctx.Value)
+	return context.WithValue(ctx, workflowTraceParametersKey{}, makeTraceabilityTags(container, entries))
+}
+
+func WithWorkflowTraceParameters(ctx workflow.Context, entries ...string) workflow.Context {
+	container := getOrCreateTags(ctx.Value)
+	return workflow.WithValue(ctx, workflowTraceParametersKey{}, makeTraceabilityTags(container, entries))
+}
+
+func getOrCreateTags(contextGetter func(key any) any) *map[string]string {
+	existing, ok := contextGetter(workflowTraceParametersKey{}).(*map[string]string)
+	return lo.IfF(ok && existing != nil, func() *map[string]string {
+		copy := map[string]string{}
+		for k, v := range *existing {
+			copy[k] = v
+		}
+		return &copy
+	}).ElseF(func() *map[string]string {
+		return &map[string]string{}
+	})
+}
+
+func makeTraceabilityTags(tags *map[string]string, entries []string) *map[string]string {
+	if len(entries)%2 != 0 {
+		panic("required key-value pair of entries")
+	}
+	for i := 0; i < len(entries); i += 2 {
+		(*tags)[entries[i]] = entries[i+1]
+	}
+	return tags
+}
+
+func TrackFinishedWorkflow(ctx workflow.Context, workflowStatus string) {
+	start := workflow.GetInfo(ctx).WorkflowStartTime
+	stop := workflow.Now(ctx)
+	duration := stop.Sub(start)
+
+	workflow.GetMetricsHandler(ctx).WithTags(map[string]string{"wfStatus": workflowStatus}).Timer("workflow_finished").Record(duration)
+}
+
+func TrackFinishedTask[T any](ctx workflow.Context, taskName string, activityRunner func() TaskResult[T]) (T, error) {
```

Comment:
Task seems Deploy Engine-centric - ideally this package would stay generic. Is there a way for this logic to live in Deploy Engine instead?

Reply:
i've modified it to be "activity generic", but it would be used by runTask in deploy-engine only (however, there is no restriction for that).